### PR TITLE
Added more feature to the library.

### PR DESCRIPTION
**Pull request by octplane**
- Several original line noise calls are now available.
- Bumped linenoise and linenoise-sys versions and dependencies
- Added sample program to test the library and demonstrate the API

### CODE REVIEW

**Comment by octplane** [comment]
NOTE THAT I HAD TO REMOVE THE MUTEX BECAUSE I NEED A RE-RENTRANT MUTEX IN ORDER TO AVOID DEADLOCKS IN COMPLETION CALLBACKS (sorry for that :-D).

**Comment by octplane** [comment]
Also, because there is an actual physical mutex (the current console), I suggest we had some caveat to the documentation and not add any exclusion code to the library...

**Comment by octplane** [comment]
@cmr Not back ?

File: src/lib.rs [hunk 1]
```diff
@@ -1,25 +1,121 @@
...
+pub fn add_completion(c: *mut linenoise::Completions, s: &str) {
+    unsafe {
+        linenoise::linenoiseAddCompletion(c, CString::from_slice(s.as_bytes()).as_slice_with_nul().as_ptr());
+    }
+}
+
+
+/// Add this string to the history
+pub fn history_add(line: &str) -> i32 {
+    let cs = CString::from_slice(line.as_bytes()).as_slice_with_nul().as_ptr();
```
Comment (emberian) [inline src/lib.rs]: The `CString` will be freed when it leaves scope, leaving a use-after-free.

File: src/lib.rs [hunk 2]
```diff
@@ -1,25 +1,121 @@
...
+                add_completion(lc, *x);
+            }
+        }
+    }
+}
+
+/// Add a completion to the current list of completions.
+pub fn add_completion(c: *mut linenoise::Completions, s: &str) {
+    unsafe {
+        linenoise::linenoiseAddCompletion(c, CString::from_slice(s.as_bytes()).as_slice_with_nul().as_ptr());
```
Comment (emberian) [inline src/lib.rs]: The `CString` will be freed when it leaves scope, leaving a use-after-free.

File: src/lib.rs [hunk 3]
```diff
@@ -1,25 +1,121 @@
...
+pub fn history_set_max_len(len: i32) -> i32 {
+    let mut ret: i32;
+    unsafe {
+        ret = linenoise::linenoiseHistorySetMaxLen(len);
+    }
+    ret
+}
+
+/// Save the history on disk
+pub fn history_save(file: &str) -> i32 {
```
Comment (emberian) [inline src/lib.rs]: This should use `c_int`.

File: src/lib.rs [hunk 4]
```diff
@@ -1,25 +1,121 @@
...
+    let fname = CString::from_slice(file.as_bytes()).as_slice_with_nul().as_ptr();
+    let mut ret: i32;
+    unsafe {
+        ret = linenoise::linenoiseHistorySave(fname);
+    }
+    ret
+}
+
+/// Load the history on disk
+pub fn history_load(file: &str) -> i32 {
```
Comment (emberian) [inline src/lib.rs]: This should use `c_int`.

File: src/lib.rs [hunk 5]
```diff
@@ -1,25 +1,121 @@
...
+}
+
+///Clears the screen
+pub fn clear_screen() {
+    unsafe {
+        linenoise::linenoiseClearScreen();
+    }
+}
+
+pub fn set_multiline(ml: i32) {
```
Comment (emberian) [inline src/lib.rs]: This should use `c_int`.

**Comment by emberian** [comment]
Sorry for the delay :)

**Comment by octplane** [comment]
I have also find another lifetime issue that was actually breaking the calls...

**Comment by octplane** [comment]
@cmr up ! :)

**Pull request merged**